### DISCUSSION TITLE
Recurring tasks with Nextcloud

### DISCUSSION THREAD

**Discussion by nilskamm** [TechbeeAT/jtxBoard#309] [Q&A] [answered]
Hey everyone! 
I´m using jtxBoard (2.02.02-rc5.ose) with Nextcloud synched by DAVx5.
I know that Nextcloud is not supporting recurring tasks at the moment and that there is no plan when this will maybe happen.
Some people told me, that some clients are able to handle recurring tasks on their own and sync it into Nextcloud with their own logic. For example Thunderbird. 
I tried the same with jtxBoard but without any success. 

So i just wanted to get it confirmed from other people. 
Is it right, that jtxBoard doesn´t handle recurring tasks with Nextcloud, so that i could work with both sites simultaneously or do i maybe just use it wrong?

Thanks a lot!!!

**Comment by patrickunterwegs** [comment, marked as answer]
Hey @nilskamm , 
from my side I can tell you that basically the sync of recurring tasks should work without problems with Nextcloud. But as you say correctly, Nextcloud does not display them (actually it just displays the main entry with the series definition). However, when making a recurring entry there is a recurring rule in the background of the entry, this should not get lost within nextcloud. So when synchronizing entries with Thunderbird and jtx Board through your Nextcloud instance, both apps should display the recurring instances correctly. (But I am not sure if Nextcloud handles the tasks correctly if you change it in their web interface). 
jtx Board 2.2.x has a bug that doesn't handle exceptions of recurring entries correctly, which would manifest in problems when synchronizing with Thunderbird. This will be fixed in jtx Board 2.3.0 (which will come within the next 2-3 weeks - or the beta hopefully within the next days). 

For sure other user input is welcome here, but can you also explain what went wrong?

**Reply by nilskamm** [reply]
I try to. 
I created two tests in jtxBoard yesterday.
1. Start: Wednesday, Due: Thursday, Prio: 1, Recurrence: Friday, Saturday, Sunday, Monday, Tuesday to only use in Nextcloud in a browser on my computer and see the outcome in jtxBoard.
2. Start: Wednesday, Due: Thursday, Prio: 1, Recurrence: Friday, Saturday, Sunday, Monday, Tuesday to only use in jtxBoard and see the outcome in Nextcloud in a browser on my computer.

From both recurring tasks only the first recurrence was listed in Nextcloud as normal tasks. No difference to other tasks i created in Nextcloud or jtxBoard. 
Differences:
1. After completing the task in Nextcloud, all recurrences and the original got completed. 
2. After completing the first recurrence in jtxBoard, nothing happend to the task in Nextcloud. After completing the second recurrence, there is still no change in Nextcloud. Nextcloud still shows the first recurrence Start: Wednesday, Due: Thursday, which is already completed in jtxBoard. 

What do you mean by "main entry with the series definition"?

**Reply by patrickunterwegs** [reply]
Alright, so I need to explain a bit: 
Basically every entry in any system that supports the iCalendar (.ics) format can be represented in an .ics file with the specific format. 
When you have a recurring entry that - let's say - repeats 10 times, the apps like Thunderbird or jtx Board would create one entry with the series definition. This definition contains the RRULE - for example repeat every Sunday 10 times. 
Apps are now responsible to interpret the series definition and handle it. jtx Board for example creates 10 additional entries in the background - the 10 instances for the series definition (the series definition would still be there in the background). 
You can also change instances which then also need special handling, but I skip that part for now.
If you change the series definition, all instances should get updated (unless the ones that have been changed).

Alright, so when you synchronize your entries from Thunderbird or jtx Board, the instances are not sent to the server (because there is no need, it is each app's responsibility to interpret the recurrence rule and show/generate the 10 entries (from the example above) - and also consider the changes that might have been made in the recurring instances. 

Nextcloud does not do the step of generating/showing the instances, so you only see the series definition. And also Nextcloud is ignoring instances and any changed instances. So if you mark the series as done in Nextcloud, that would mean that actually all instances should be marked as done (but you can't see that, as you only see the series definition). But when you synchronize with jtx Board or Thunderbird, you should get all instances marked as done (as the series element was marked as done).

The problem here is really the Nextcloud web UI that doesn't support recurring entries. Even worse, it shows the series definition as a standalone task without giving any hint that there is a series defined in the background. In my opinion Nextcloud should either support series properly or at least just not allow editing of series and show a proper message for the user...

**Reply by nilskamm** [reply]
Thanks a lot for this very detailed answer! 
So i´m waiting for Nextcloud to be able to handle it well someday and will just not have recurring tasks until then.

**Reply by patrickunterwegs** [reply]
Yes, I think there's no point in using the web UI of Nextcloud with recurring entries. You could use Thunderbird and jtx Board though (and just not use the web UI of Nextcloud). Just better wait for jtx Board 2.3.x as the recurrence handling got some major improvements there (especially concerning the compatibility with Thunderbird) 😉

**Reply by nilskamm** [reply]
Yeah... sadly that´s not an option. I work on several Nextcloud instances with a bunch of people with the same tasks. So it has to be something that just works. And just working with all people at the moment is to create "recurring" tasks by hand. But thank you for everything!

**Reply by patrickunterwegs** [reply]
Well, for this case I still have good news for you 😉 
As I said jtx Board 2.3.0 will have a refactored recurrence handling. One feature request was also to have subtasks for recurring entries. But this is technically not possible for the reason that I explained above - as the recurring instances are just generated (and don't have their own identifier), they cannot have a linked subtask. 
So I searched for a compromise: I added a button that allows you to transform all instances of a series to standalone entries (the series definition would be deleted in the background). That would then be fully compatible with Nextcloud without the need to create every single entry 😉   (the series definition will be missing, so you can't update all of them at once anymore, but you could still easily create another series and transform them into standalone entries again)!
